File: packages/android/src/main/java/com/kozmos/components/AISearchButton/AISearchButton.kt
package com.kozmos.components.aisearchbutton

import android.provider.Settings
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.LocalContext
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosColors

/**
 * The AI search, beside the search field: a 43 disc inside a 48 ring whose
 * sweep gradient runs through the theme's own ramp — 300 to 600 and back, the
 * first gradient the system draws, made of tokens — with a 16 icon. The
 * button shows the icon alone and is named by its label.
 */
@Composable
fun KozmosAISearchButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    label: String = "AI search",
    enabled: Boolean = true
) {
    // The prototype's: a 48 circle whose gradient ring is a band two and a
    // half wide around a 43 white disc — nothing overflows the row.
    Box(
        modifier = modifier
            .size(48.dp)
            .clickable(enabled = enabled, role = Role.Button, onClick = onClick)
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center
    ) {
        // The gradient turns in place, the prototype's 3.6 seconds a turn, unless the
        // system's animations are off; the first frame is the resting one.
        val transition = rememberInfiniteTransition(label = "kozmos-ai-search-ring")
        val ringAngle by transition.animateFloat(
            initialValue = 0f,
            targetValue = 360f,
            animationSpec = infiniteRepeatable(animation = tween(durationMillis = 3600, easing = LinearEasing), repeatMode = RepeatMode.Restart),
            label = "kozmos-ai-search-ring-angle"
        )
        val context = LocalContext.current
        val animationsOn = remember(context) {
            Settings.Global.getFloat(context.contentResolver, Settings.Global.ANIMATOR_DURATION_SCALE, 1f) > 0f
        }
        Box(
            modifier = Modifier
                .size(48.dp)
                .graphicsLayer { rotationZ = if (animationsOn) ringAngle else 0f }
                .clip(CircleShape)
                // The prototype's rainbow from the system's own data colours
                // and its success green.
                .background(
                    Brush.sweepGradient(
                        listOf(
                            KozmosColors.semanticsDataRed,
                            KozmosColors.semanticsDataYellow,
                            KozmosColors.primitivesColorsEmotionalSuccess500,
                            KozmosColors.semanticsDataTeal,
                            KozmosColors.semanticsDataBlue,
                            KozmosColors.semanticsDataPurple,
                            KozmosColors.semanticsDataRed
                        )
                    )
                )
        )
        Surface(
            modifier = Modifier.size(43.dp),
            shape = CircleShape,
            color = KozmosColors.primitivesColorsBackground0
        ) {
            Box(contentAlignment = Alignment.Center) {
                Icon(
                    imageVector = Icons.Default.AutoAwesome,
                    contentDescription = null,
                    tint = KozmosColors.primitivesColorsTheme500,
                    modifier = Modifier.size(16.dp)
                )
            }
        }
    }
}

File: packages/ios/Sources/Components/AISearchButton/AISearchButton.swift
import SwiftUI

/// The AI search, beside the search field: a 44 disc inside a 66 ring whose
/// gradient runs through the theme's own ramp — 300 to 600 and back, the
/// first gradient the system draws, made of tokens — with a 16 icon. The
/// button shows the icon alone and is named by its label.
public struct KozmosAISearchButton: View {
    let label: String
    let action: () -> Void

    public init(label: String = "AI search", action: @escaping () -> Void) {
        self.label = label
        self.action = action
    }

    static let ring: CGFloat = 66
    static let disc: CGFloat = 44
    /// The button's own footprint, the prototype's 48: the ring overflows it,
    /// so a row that holds the field and this button stays the field's height.
    static let footprint: CGFloat = 48

    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    /// The ring's gradient turns in place, once every three seconds, unless
    /// motion is reduced; the first frame is the resting one.
    @State private var ringAngle: Double = 0
    static let spinDuration: Double = 3

    public var body: some View {
        Button(action: action) {
            ZStack {
                Circle()
                    .fill(KozmosColors.primitivesColorsBackground0)
                    .frame(width: Self.disc, height: Self.disc)
                    .kozmosElevation(KozmosShadows.semanticsElevationRaised)
                Image(systemName: "sparkles")
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundColor(KozmosColors.primitivesColorsTheme500)
                    .accessibilityHidden(true)
            }
            .frame(width: Self.footprint, height: Self.footprint)
            // Drawn, not laid out: the ring takes no room from the row.
            .background(
                Circle()
                    .fill(
                        AngularGradient(
                            colors: [
                                KozmosColors.primitivesColorsTheme300,
                                KozmosColors.primitivesColorsTheme600,
                                KozmosColors.primitivesColorsTheme300,
                            ],
                            center: .center
                        )
                    )
                    .frame(width: Self.ring, height: Self.ring)
                    // The gradient turns; the circle it fills does not move.
                    .rotationEffect(.degrees(ringAngle))
            )
            .contentShape(Circle())
        }
        .onAppear {
            guard !reduceMotion else { return }
            withAnimation(.linear(duration: Self.spinDuration).repeatForever(autoreverses: false)) {
                ringAngle = 360
            }
        }
        .buttonStyle(.plain)
        .accessibilityLabel(label)
    }
}
